File: benchmark/bench.jl
using BenchmarkTools

include("../src/Simplex.jl")
using .Simplex

"""Generate a random LP with m constraints and n total variables (n > m)."""
function random_lp(m::Int, n::Int; seed::Int=42)
    rng = MersenneTwister(seed)
    A_raw = rand(rng, m, n - m)
    A = [A_raw Matrix{Float64}(I, m, m)]   # append slack columns
    b = abs.(rand(rng, m)) .+ 0.1            # strictly positive RHS
    c = [rand(rng, n - m); zeros(m)]         # only original vars in objective
    LinearProgram(A, b, c)
end

"""Build benchmark random suite"""
function build_benchmark_suite()
    suite = BenchmarkGroup()

    # Small canonical problem (used in tests)
    lp_small = LinearProgram(
        Float64[2 1 1 0; 1 2 0 1],
        Float64[4; 4],
        Float64[4; 3; 0; 0])
    basis_small = BasisState([3, 4], [1, 2])

    suite["small/phase2_only"] = @benchmarkable solve_phase2($lp_small, $basis_small)
    suite["small/two_phase"] = @benchmarkable solve($lp_small)

    # Medium random LPs
    for (m, n) in [(10, 20), (50, 100), (100, 200)]
        lp = random_lp(m, n)
        tag = "m=$(m)_n=$(n)"
        suite["random/$tag"] = @benchmarkable solve($lp)
    end
    return suite
end

function run_benchmark_suite(suite)
    println("Tuning benchmarks…")
    tune!(suite)
    println("Running benchmarks…")
    results = run(suite, verbose=true)

    println("\n=== Results ===")
    for (name, trial) in leaves(results)
        println("$name: median=$(BenchmarkTools.median(trial))")
    end
end

run_benchmark_suite(build_benchmark_suite())

File: python/benchmark/bench.py
from __future__ import annotations

import time

import numpy as np

from simplex import BasisState, LinearProgram, solve, solve_phase2

N_RUNS = 10


def random_lp(m: int, n: int, seed: int = 42) -> LinearProgram:
    rng   = np.random.default_rng(seed)
    A_raw = rng.random((m, n - m))
    A     = np.hstack([A_raw, np.eye(m)])
    b     = np.abs(rng.random(m)) + 0.1
    c     = np.concatenate([rng.random(n - m), np.zeros(m)])
    return LinearProgram(A, b, c)


def benchmark(name: str, fn, n_runs: int = N_RUNS) -> None:
    times = []
    for _ in range(n_runs):
        t0 = time.perf_counter()
        fn()
        times.append(time.perf_counter() - t0)
    median_ms = np.median(times) * 1e3
    print(f"{name}: median={median_ms:.3f} ms")


def main() -> None:
    lp_small    = LinearProgram(
        np.array([[2, 1, 1, 0], [1, 2, 0, 1]], dtype=float),
        np.array([4, 4], dtype=float),
        np.array([4, 3, 0, 0], dtype=float),
    )
    basis_small = BasisState(basic=[2, 3], nonbasic=[0, 1])

    benchmark("small/phase2_only", lambda: solve_phase2(lp_small, basis_small))
    benchmark("small/two_phase",   lambda: solve(lp_small))

    for m, n in [(10, 20), (50, 100), (100, 200)]:
        lp  = random_lp(m, n)
        tag = f"m={m}_n={n}"
        benchmark(f"random/{tag}", lambda lp=lp: solve(lp))


if __name__ == "__main__":
    main()
